libs/extended/src/Network/HTTP/Client/Extended.hs

```diff
@@ -0,0 +1,101 @@
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+-- | A version of "Network.HTTP.Client" with extra utilities.
+module Network.HTTP.Client.Extended
+    ( module Network.HTTP.Client
+    -- * Connection limits
+    , addPerHostConnectionLimit
+    , PerHostConnectionLimitReached(..)
+    ) where
+
+import Imports
+import Control.Exception (throwIO)
+import Control.Lens
+import Data.ByteString.Short
+import Data.Hashable
+import Network.HTTP.Client
+import Network.HTTP.Client.Internal (Connection(..), ManagerSettings(..))
+
+import qualified STMContainers.Map as STMMap
+
+-- | Connection destination: just host and port.
+data Destination = Destination {-# UNPACK #-} !ShortByteString !Int
+    deriving (Eq, Show, Generic)
+
+instance Hashable Destination
+
+-- | Add a per-host hard connection limit. When there are more than N open connections to a
+-- host, any new connection to that host will immediately fail with 'PerHostConnectionLimit'.
```

Comment:
I assume this is per-DNS-name, so it may be a bunch of hosts behind a load balancer?  May be worth clarifying the comment.

What if it's more complicated, and we want to have a limit per all hosts satisfying a certain property?  (I assume that's not required for our use case, in that case we shouldn't implement it now.)

Reply:
If we need a more complicated rule, we can pass a function from host and port to some Hashable a.